dask/array/numpy_compat.py

```diff
@@ -359,3 +361,27 @@ def nancumprod(a, axis=None, dtype=None, out=None):
         """
         a, mask = _replace_nan(a, 1)
         return np.cumprod(a, axis=axis, dtype=dtype, out=out)
+
+
+def _make_sliced_dtype(dtype, index):
+    if LooseVersion(np.__version__) >= LooseVersion("1.14.0"):
+        return _make_sliced_dtype_new(dtype, index)
+    else:
+        return _make_sliced_dtype_old(dtype, index)
+
+
+def _make_sliced_dtype_new(dtype, index):
+    # For https://github.com/numpy/numpy/pull/6053
+    # TODO: handle either positional or named indexing
+    new = {
+        'names': index,
+        'formats': [dtype.fields[name][0] for name in index],
+        'offsets': [dtype.fields[name][1] for name in index],
+        'itemsize': dtype.itemsize,  # is this true?
```

Comment:
Yes this is right. The new dtype should be the just like the old dtype except with some fields removed. The new dtype has to have the same itemsize since it is used to view the old data.

Reply:
Note that this code, as well as your old code, ignores field titles. 

I'm not sure if its worth the effort to support them, but I know some numpy users use them. We recently got a bug report about indexing involving titles which is now fixed: https://github.com/numpy/numpy/issues/9625

You'd have to add an extra `titles` key here. I probably wouldn't worry about it for the old code, since because of that bug titles didn't index properly before numpy 1.14.

Reply:
@ahaldane with that, the `np.dtype` constructor complains about `ValueError: title already used as a name or title.`

I'm not really sure if that should work or not, but I suspect the intent is for the dtype repr to be evalable into a dtype? Should I make an issue on NumPy?

```python
In [50]: a = np.zeros(4, dtype=[(('a', 'b'), 'i'), ('c', 'i'), ('d', 'i')])

In [51]: a[['a']].dtype
Out[51]: dtype({'names':['a'], 'formats':['<i4'], 'offsets':[0], 'titles':['a'], 'itemsize':12})

In [52]: np.dtype({'names':['a'], 'formats':['<i4'], 'offsets':[0], 'titles':['a'], 'itemsize':12})
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-52-9f350bf6f22d> in <module>()
----> 1 np.dtype({'names':['a'], 'formats':['<i4'], 'offsets':[0], 'titles':['a'], 'itemsize':12})

ValueError: title already used as a name or title.
```

Reply:
Yeah if it's giving you trouble I would ignore titles for now. Some devs have thought we should deprecate them of caution against them, but enough users still want them that we support them.

As to your exception.. titles cannot be the same string as a field name; you can't have both a title and a name be 'a'. They need to be different, so your example should be
```
np.dtype({'names':['a'], 'formats':['<i4'], 'offsets':[0], 'titles':['b'], 'itemsize':12})
```

Reply:
Wait I just reread you comment. I don't get the same as you for the line `a[['a']].dtype`. In numpy 1.13, for both py3/2, that returns `dtype([('a', '<i4')])` here, and it raises an exception in numpy master.

Not sure if you want to try debugging this since it might not be worth the effort, but if you could double check that line and tell me your numpy version, I'll investigate.

Reply:
I was just reading through the docs and getting very confused why my example didn't raise :) I was on the merge commit for https://github.com/numpy/numpy/pull/6053 after doing a git bisect. Trying it out on master now.

Reply:
All good, `Out[51]` (correctly) raises on NumPy master for me.